### PR TITLE
Locked jobs

### PR DESCRIPTION
**Pull request by ggueret**
- Added 'expired' status for the jobs who were unable to lock
- Added 'locked_job' bool option to `mrq.task.Task`, to lock a job execution by his task path.
- Mark `mrq.basetasks.cleaning.[RequeueInterruptedJobs, RequeueRetryJobs, RequeueStartedJobs, RequeueRedisStartedJobs, RequeueLostJobs, MigrateKnownQueues, CleanKnownQueues, ]`  as locked jobs.

### CODE REVIEW

**Comment by sylvinus** [comment]
- `locked_job` is probably not the best name
 - `lock_timeout` is not explicit enough (this lock should have a specific name to avoid confusion)
 - The locking code is not safe enough
 - I'm not (yet) convinced we need to add a new status `expired` just for this feature.

File: mrq/basetasks/cleaning.py [hunk 1]
```diff
@@ -10,6 +10,8 @@ class RequeueInterruptedJobs(Task):
 
     """ Requeue jobs that were marked as status=interrupt when a worker got a SIGTERM. """
 
+    max_concurrency = 1
```
Comment (sylvinus) [inline mrq/basetasks/cleaning.py]: "maxconcurrency" serait plus facile à grepper ?

File: mrq/config.py [hunk 2]
```diff
@@ -391,6 +391,13 @@ def add_parser_args(parser, config_type):
             help='Strategy for dequeuing multiple queues. Default is \'sequential\',' +
                  'to dequeue them in command-line order.')
 
+        parser.add_argument(
+            '--lock_timeout',
```
Comment (sylvinus) [inline mrq/config.py]: --maxconcurrency_timeout

File: mrq/config.py [hunk 3]
```diff
@@ -391,6 +391,13 @@ def add_parser_args(parser, config_type):
...
+            default=1.,
+            type=float,
+            action='store',
+            help='Max seconds to wait for a lock before changing the job status to expired.')
```
Comment (sylvinus) [inline mrq/config.py]: text to update

File: mrq/job.py [hunk 4]
```diff
@@ -274,7 +279,22 @@ def perform(self):
 
         self.task.is_main_task = True
 
-        result = self.task.run_wrapped(self.data["params"])
+        try:
+            if self.task.max_concurrency:
+                pipeline = context.connections.redis.pipeline()
+                pipeline.incr(self.redis_concurrency_key)
+                pipeline.expireat(self.redis_concurrency_key, int(time.time()) + self.timeout)
+                current = pipeline.execute()[0]
```
Comment (sylvinus) [inline mrq/job.py]: this is not safe enough - with maxconcurrency=2, if a worker dies here, the count will stay at a minimum at 1, and if there are still tasks happening, the key will never expire.

File: mrq/basetasks/cleaning.py (hunk 1 above)
Comment (ggueret) [inline mrq/basetasks/cleaning.py]: c'est par respect de la sémantique existante, avec par exemple le `Job.max_retries`

**Comment by ggueret** [comment]
- Added 'maxconcurrency' status for the jobs who reach the max_concurrency limit.
- Replaced 'locked_job' by 'max_concurrency' who now take an integer value.
- Mark mrq.basetasks.cleaning.[RequeueInterruptedJobs, RequeueRetryJobs, RequeueStartedJobs, RequeueRedisStartedJobs, RequeueLostJobs, MigrateKnownQueues, CleanKnownQueues] with a concurrency of 1 per `Job.timeout`.

**Pull request merged**